Track source location in change detection

aevyrie opened the pull request:
# Objective

- Make it possible to know *what* changed your component or resource.
- Common need when debugging, when you want to know the last code location that mutated a value in the ECS.
- This feature would be very useful for the editor alongside system stepping.

## Solution

- Adds the caller location to column data.
- Mutations now `track_caller` all the way up to the public API.
- Commands that invoke these functions immediately call `Location::caller`, and pass this into the functions, instead of the functions themselves attempting to get the caller. This would not work for commands which are deferred, as the commands are executed by the scheduler, not the user's code.

## Testing

- The `component_change_detection` example now shows where the component was mutated: 

```
2024-07-28T06:57:48.946022Z  INFO component_change_detection: Entity { index: 1, generation: 1 }: New value: MyComponent(0.0)
2024-07-28T06:57:49.004371Z  INFO component_change_detection: Entity { index: 1, generation: 1 }: New value: MyComponent(1.0)
2024-07-28T06:57:49.012738Z  WARN component_change_detection: Change detected!
        -> value: Ref(MyComponent(1.0))
        -> added: false
        -> changed: true
        -> changed by: examples/ecs/component_change_detection.rs:36:23
```

- It's also possible to inspect change location from a debugger:
<img width="608" alt="image" src="https://github.com/user-attachments/assets/c90ecc7a-0462-457a-80ae-42e7f5d346b4">


---

## Changelog

- Added source locations to ECS change detection behind the `track_change_detection` flag.

## Migration Guide

- Added `changed_by` field to many internal ECS functions used with change detection when the `track_change_detection` feature flag is enabled. Use Location::caller() to provide the source of the function call.

cart commented:
`Mut` is one of the hottest paths in Bevy. I'm concerned about the performance implications of introducing _anything_ (even the cheapest ops) to it. I'd like to either see compelling evidence that this won't meaningfully affect performance, especially given that this is a "debugging tool". I'd feel much more comfortable if this was behind a feature flag.

alice-i-cecile commented:
I really like this feature and think it'll be very useful, but I'll second the request for putting it behind an off-by-default feature flag.

aevyrie commented:
Yup, I assumed as much. I was hoping to get a read on whether this is even an acceptable approach to take for the feature.

BD103 commented:
> Is there a reason you're de referencing the static reference you get from [`Location::caller`](https://doc.rust-lang.org/std/panic/struct.Location.html#method.caller)? Couldn't you just pass the reference around?

`Location` is two `u32`s and one `&str`, making it 16 bytes on 64-bit platforms. Storing a reference instead would bring it down to 8 bytes, which might have some mild performance gains because it would fit in a register.

Furthermore, why do you use `core::panic::Location` as compared with the `std` variant?

aevyrie commented:
static reference: I started with this, but was having trouble with the deref making it hard to update the field, so I switched to an owned value to get it working. I agree it should be fixed if this is going to be merged.
using core: because it can't hurt? I wasn't sure if bevy_ecs was every intended to be used in a no_std context, and figured it would be safer to use core for a type so ubiquitous.

BD103 commented:
> 2\. Changes from commands completely break this approach because commands are deferred, and the call stack can't be traced to the calling location by the time the command is actually executed.

Is it possible to find the `Location` in `Commands::insert`, then pass that forward to the actual command?

alice-i-cecile commented:
> using core: because it can't hurt? I wasn't sure if bevy_ecs was ever intended to be used in a no_std context, and figured it would be safer to use core for a type so ubiquitous.

I would like this, one day :) Probably core + alloc, but still! Not a priority, but might as well do it when there's no cost to it.

aevyrie commented:
Good call! That looks pretty straightforward.

slyedoc commented:
Spent far to long trying to figure out why a camera transform from glb was getting reset, was correct on spawn, next frame its Identity, this found the issue in seconds.  Was calling remove_parent_in_place before TransformPropagate had ran.  Huge help.

BD103 commented:
This is close to finished, it just needs a system to track the caller from `Commands`. That's past my capabilities, but hopefully someone else is up for the task!

BD103 commented on crates/bevy_ecs/src/world/mod.rs:
```diff
@@ -1268,7 +1268,10 @@ impl World {
+    #[track_caller]
     pub fn init_resource<R: Resource + FromWorld>(&mut self) -> ComponentId {
+        #[cfg(feature = "track_change_detection")]
+        let caller = core::panic::Location::caller();
```
```suggestion
        let caller = Location::caller();
```

`Location` is imported, you don't need to qualified syntax :)

BD103 commented on crates/bevy_ecs/src/world/mod.rs:
```diff
@@ -1293,12 +1301,20 @@ impl World {
+    #[track_caller]
     pub fn insert_resource<R: Resource>(&mut self, value: R) {
+        #[cfg(feature = "track_change_detection")]
+        let caller = core::panic::Location::caller();
```
```suggestion
        let caller = Location::caller();
```

BD103 commented on crates/bevy_ecs/src/world/mod.rs:
```diff
@@ -1316,6 +1332,8 @@ impl World {
     #[inline]
     pub fn init_non_send_resource<R: 'static + FromWorld>(&mut self) -> ComponentId {
+        #[cfg(feature = "track_change_detection")]
+        let caller = core::panic::Location::caller();
```
```suggestion
        let caller = Location::caller();
```

BD103 commented on crates/bevy_ecs/src/world/mod.rs:
```diff
@@ -1344,12 +1367,20 @@ impl World {
+    #[track_caller]
     pub fn insert_non_send_resource<R: 'static>(&mut self, value: R) {
+        #[cfg(feature = "track_change_detection")]
+        let caller = core::panic::Location::caller();
```
```suggestion
        let caller = Location::caller();
```

BD103 commented on crates/bevy_ecs/src/world/mod.rs:
```diff
@@ -1598,6 +1629,8 @@ impl World {
         func: impl FnOnce() -> R,
     ) -> Mut<'_, R> {
+        #[cfg(feature = "track_change_detection")]
+        let caller = core::panic::Location::caller();
```
```suggestion
        let caller = Location::caller();
```

BD103 commented on crates/bevy_ecs/src/world/mod.rs:
```diff
@@ -1988,17 +2074,24 @@ impl World {
         &mut self,
         component_id: ComponentId,
         value: OwningPtr<'_>,
+        #[cfg(feature = "track_change_detection")] caller: &'static core::panic::Location,
```
```suggestion
        #[cfg(feature = "track_change_detection")] caller: &'static Location,
```

BD103 commented on crates/bevy_ecs/src/world/mod.rs:
```diff
@@ -1961,17 +2040,24 @@ impl World {
         &mut self,
         component_id: ComponentId,
         value: OwningPtr<'_>,
+        #[cfg(feature = "track_change_detection")] caller: &'static core::panic::Location,
```
```suggestion
        #[cfg(feature = "track_change_detection")] caller: &'static Location,
```

BD103 reviewed (approved):
Let's go, thank you! 🎉

aevyrie commented:
I've updated this to now work with commands, as well as a pass on naming, and updated the change detection example to more effectively teach these concepts.

BD103 commented on crates/bevy_ecs/src/world/mod.rs:
```diff
@@ -1303,8 +1311,21 @@ impl World {
+        );
+    }
+
+    /// Split into a new function so we can pass the calling location into the function when using
+    /// as a command.
```
```suggestion
    // Split into a new function so we can pass the calling location into the function when using
    // as a command.
```

(Non-blocking) These comments are more notes than documentation, so I think `//` works better in this case.

aevyrie replied:
I don't see why you wouldn't want to have this show up in the docs/IDE. It's public only to the crate, so the only people seeing this will be users of bevy_ecs, in which case this is useful context when deciding which function to use.

BD103 commented on examples/ecs/change_detection.rs:
```diff
@@ -0,0 +1,106 @@
+    // Note the first change detection log correctly points to this line because the component is
+    // added. Although commands are deferred, they are able to track the original calling location.
+    commands.spawn(MyComponent(0.0));
+    commands.insert_resource(MyResource(0.0))
```
```suggestion
    commands.insert_resource(MyResource(0.0));
```

Clippy wants a semicolon here, but that should make CI pass.

NthTensor reviewed (approved):
I like this and I want it. Should `Commands::add` and `Command::push` be `track_caller` so that custom user commands can provide call tracing? Or would that be something user should do on the methods of their extension trait.

BD103 commented:
> I like this and I want it. Should `Commands::add` and `Command::push` be `track_caller` so that custom user commands can provide call tracing? Or would that be something user should do on the methods of their extension trait.

If we do want to support custom commands like this, it would need more documentation. I feel like this is probably follow-up PR territory.

NthTensor commented:
> follow-up PR territory.

Fair enough, agreed.

MrGVSV commented on examples/ecs/change_detection.rs:
```diff
@@ -0,0 +1,106 @@
+            component.is_added(),
+            component.is_changed(),
+            // If you enable the `track_change_detection` feature, you can unlock the `changed_by()`
+            // method. It returns the the file and line number that the component or resource was
```
```suggestion
            // method. It returns the file and line number that the component or resource was
```

workingjubilee commented:
> > Is there a reason you're de referencing the static reference you get from [`Location::caller`](https://doc.rust-lang.org/std/panic/struct.Location.html#method.caller)? Couldn't you just pass the reference around?
> 
> `Location` is two `u32`s and one `&str`, making it 16 bytes on 64-bit platforms. Storing a reference instead would bring it down to 8 bytes, which might have some mild performance gains because it would fit in a register.
> 

at the risk of sounding incredibly pedantic: 24 bytes. `&str` is the pointer and the length, so it is already 16 bytes.

remember that this is a pretty magical struct that gets willed into existence by the compiler during codegen.

aevyrie commented:
> insert_with_caller & friends aren't pub. So the actual question is: Do we want to make them pub?

I think so. This will be needed for custom impls, as you said. We can document the context around this on the `with_caller` methods. I wish there was a way to make this pattern nicer, but alas.

BD103 commented:
> at the risk of sounding incredibly pedantic: 24 bytes. `&str` is the pointer and the length, so it is already 16 bytes.

Hah, I forgot about wide pointers. Thanks!

alice-i-cecile commented:
Thank you to everyone involved with the authoring or reviewing of this PR! This work is relatively important and needs release notes! Head over to https://github.com/bevyengine/bevy-website/issues/1666 if you'd like to help out.
